src/service/cid-api.js

```diff
@@ -0,0 +1,172 @@
+};
+
+const TAG = 'GoogleCidApi';
+const AMP_TOKEN = 'AMP_TOKEN';
+
+const TokenStatus = {
+  RETRIEVING: '$RETRIEVING',
+  OPT_OUT: '$OPT_OUT',
+  ERROR: '$ERROR',
+};
+
+const TIMEOUT = 30000;
+const DAY = 24 * 60 * 60 * 1000;
+const YEAR = 365 * DAY;
+
+/**
+ * Client impl for Google CID API
+ */
+export class GoogleCidApi {
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    this.win_ = win;
+    /**
+     * @private {!./timer-impl.Timer}
+     */
+    this.timer_ = Services.timerFor(this.win_);
+
+    /**
+     * @private {!Object<string, !Promise<?string>>}
+     */
+    this.cidPromise_ = {};
+  }
+
+  /**
+   * @param {string} scope
+   * @param {string} apiClient
+   * @return {!Promise<?string>}
+   */
+  getScopedCid(scope, apiClient) {
+    const url = this.getUrl_(apiClient);
+    if (!url) {
+      return Promise.resolve(/** @type {?string} */(null));
+    }
+
+    if (this.cidPromise_[scope]) {
+      return this.cidPromise_[scope];
+    }
+    let token;
+    // Block the request if a previous request is on flight
+    // Poll every 200ms. Longer interval means longer latency for the 2nd CID.
+    return this.cidPromise_[scope] = this.timer_.poll(200, () => {
+      token = getCookie(this.win_, AMP_TOKEN);
+      return token !== TokenStatus.RETRIEVING;
+    }).then(() => {
+      if (token === TokenStatus.OPT_OUT || token === TokenStatus.ERROR) {
+        return null;
+      }
+
+      if (!token) {
+        this.persistToken_(TokenStatus.RETRIEVING, TIMEOUT);
+      }
+      return this.fetchCid_(dev().assertString(url), scope, token)
+          .then(this.handleResponse_.bind(this, scope))
+          .catch(e => {
+            this.persistToken_(TokenStatus.ERROR, TIMEOUT);
+            dev().error(TAG, e);
+            return null;
+          });
+    });
+  }
+
+  /**
+   * @param {string} url
+   * @param {string} scope
+   * @param {?string} token
+   * @return {!Promise<!JsonObject>}
+   */
+  fetchCid_(url, scope, token) {
+    const payload = dict({
+      'originScope': scope,
+    });
+    if (token) {
+      payload['securityToken'] = token;
+    }
+    return this.timer_.timeoutPromise(
+        TIMEOUT,
+        Services.xhrFor(this.win_).fetchJson(url, {
+          method: 'POST',
+          ampCors: false,
+          credentials: 'include',
+          mode: 'cors',
+          body: payload,
+        }).then(res => res.json()));
+  }
+
+  /**
+   * @param {string} scope
+   * @param {!JsonObject} res
+   * @return {?string}
+   */
+  handleResponse_(scope, res) {
+    if (res['optOut']) {
+      this.persistToken_(TokenStatus.OPT_OUT, YEAR);
+      return null;
+    }
+    if (res['clientId']) {
+      this.persistToken_(res['securityToken'], YEAR);
+      setCookie(this.win_, scope, res['clientId'], this.expiresIn_(YEAR));
+      return res['clientId'];
+    } else {
+      this.persistToken_(TokenStatus.ERROR, DAY);
+      return null;
```

Comment:
Lets fall back to the existing cookie value if there is one, not just return null.